File: leetcode_75/Stack/python/735.py
from collections import deque
from typing import List


class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        stack = deque()
        for ast in asteroids:
            cur = ast
            stack.append(cur)
            if len(stack) >= 2:
                prev = stack[-2]
                while cur < 0 and prev > 0:
                    if abs(cur) > abs(prev):
                        stack[-2] = cur
                        stack.pop()
                    elif abs(cur) == abs(prev):
                        stack.pop()
                        stack.pop()
                    else:
                        stack.pop()

                    if len(stack) < 2:
                        break

                    cur = stack[-1]
                    prev = stack[-2]

        return list(stack)

File: leetcode_75/Stack/cpp/735.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution
{
public:
    vector<int> asteroidCollision(vector<int> &asteroids)
    {
        vector<int> stack;
        for (auto ast : asteroids)
        {
            int cur = ast;
            stack.push_back(cur);
            if (stack.size() >= 2)
            {
                int prev = stack[stack.size() - 2];
                while (cur < 0 && prev > 0)
                {
                    if (abs(cur) > abs(prev))
                    {
                        stack[stack.size() - 2] = cur;
                        stack.pop_back();
                    }
                    else if (abs(cur) == (abs(prev)))
                    {
                        stack.pop_back();
                        stack.pop_back();
                    }
                    else
                    {
                        stack.pop_back();
                    }

                    if (stack.size() < 2)
                    {
                        break;
                    }

                    cur = stack[stack.size() - 1];
                    prev = stack[stack.size() - 2];
                }
            }
        }

        return stack;
    }
};